src/ophyd_async/epics/pmac/_utils.py

```diff
@@ -40,63 +178,112 @@ def from_slice(cls, slice: Slice[Motor], ramp_up_time: float) -> _Trajectory:
+        segments: list[FillableWindow] = []
+        total_gap_points = 0
+        for collection_start, collection_end in zip(
+            collection_windows[:-1], collection_windows[1:], strict=False
+        ):
+            # Add a collection window that we will fill later
+            segments.append(
+                CollectionWindow(
+                    start=collection_start,
+                    end=collection_end,
+                    slice=slice,
+                    half_durations=half_durations,
+                )
+            )
+
+            # Now we precompute our gaps
+            # Get entry velocities, exit velocities, and distances across gap
+            entry_velocities, exit_velocities, distances = (
+                _get_entry_and_exit_velocities(
+                    motors, slice, half_durations, collection_end
+                )
+            )
+
+            # Get velocity and time profiles across gap
+            time_arrays, velocity_arrays = _get_velocity_profile(
+                motors, motor_info, entry_velocities, exit_velocities, distances
+            )
+
+            gap_segment = _calculate_profile_from_velocities(
+                motors,
+                slice,
+                collection_end,
+                time_arrays,
+                velocity_arrays,
+            )
+
+            # Add a gap segment
+            segments.append(gap_segment)
+
+            total_gap_points += gap_segment.gap_length
+
+        # "collection_windows" does not include final window,
+        # as no subsequeny gap to mark its termination
+        # So, we add it here, ending it at the end of the slice
+        segments.append(
+            CollectionWindow(collection_windows[-1], len(slice), slice, half_durations)
+        )
+
         positions: dict[Motor, npt.NDArray[np.float64]] = {}
         velocities: dict[Motor, npt.NDArray[np.float64]] = {}
 
         # Initialise arrays
-        positions = {motor: np.empty(2 * scan_size + 1, float) for motor in motors}
-        velocities = {motor: np.empty(2 * scan_size + 1, float) for motor in motors}
-        durations: npt.NDArray[np.float64] = np.empty(2 * scan_size + 1, float)
-        user_programs: npt.NDArray[np.int32] = np.ones(2 * scan_size + 1, float)
-        user_programs[-1] = 8
+        # Trajectory size calculated from 2 points per frame (midpoint and upper)
+        # Plus an initial lower point
+        # Plus PVT points added between collection windows (gaps)
+        trajectory_size = ((2 * scan_size) + 1) + total_gap_points
+        positions = {motor: np.empty(trajectory_size, float) for motor in motors}
+        velocities = {motor: np.empty(trajectory_size, float) for motor in motors}
+        durations: npt.NDArray[np.float64] = np.empty(trajectory_size, float)
+        # Default to program 1 to assume we acquire at all points
+        user_programs: npt.NDArray[np.int32] = np.empty(trajectory_size, float)
```

Comment:
Could we create the `_Trajectory` instance here and pass it into `insert_points_into_trajectory` rather than passing the four separate containers?